File: Excel/1000-1099/1031/Challenge 1031.R
library(tidyverse)
library(readxl)

path <- "1000-1099/1031/1031 Room Tariff.xlsx"
rooms <- read_excel(path, range = "A2:D22")
expected <- read_excel(path, range = "F2:G6")

dec_start <- as.Date("2026-12-01")
dec_end <- as.Date("2027-01-01")
peak_start <- as.Date("2026-12-26")

result <- rooms %>%
  mutate(
    Check_In = as.Date(Check_In),
    Check_Out = as.Date(Check_Out),
    out = coalesce(Check_Out, dec_end),
    from = if_else(Check_In < dec_start, dec_start, Check_In),
    to = if_else(out > dec_end, dec_end, out),
    nights = pmax(0, as.numeric(to - from)),
    full_december = Check_In <= dec_start &
      (is.na(Check_Out) | Check_Out >= dec_end),
    surcharge_nights = pmax(
      0,
      as.numeric(to - if_else(from < peak_start, peak_start, from))
    ),
    surcharge_nights = if_else(full_december, 0, surcharge_nights),
    revenue = nights * Base_Rate + surcharge_nights * 50
  ) %>%
  summarise(Dec_Revenue = sum(revenue), .by = Room_Type) %>%
  arrange(Room_Type)

print(result)
print(all.equal(result, expected))

File: Excel/1000-1099/1031/Challenge 1031.py
import pandas as pd

p = "1000-1099/1031/1031 Room Tariff.xlsx"
rooms = pd.read_excel(p, usecols="A:D", skiprows=1, nrows=20)
expected = pd.read_excel(p, usecols="F:G", skiprows=1, nrows=4)
expected.columns = ["Room_Type", "Dec_Revenue"]

dec_start, dec_end = pd.Timestamp("2026-12-01"), pd.Timestamp("2027-01-01")
peak_start = pd.Timestamp("2026-12-26")

rooms["out"] = rooms["Check_Out"].fillna(dec_end)
rooms["from"] = rooms["Check_In"].clip(lower=dec_start)
rooms["to"] = rooms["out"].clip(upper=dec_end)
rooms["nights"] = (rooms["to"] - rooms["from"]).dt.days.clip(lower=0)
rooms["full_december"] = (rooms["Check_In"] <= dec_start) & (
    rooms["Check_Out"].isna() | (rooms["Check_Out"] >= dec_end)
)
rooms["surcharge_nights"] = (
    rooms["to"].clip(lower=peak_start) - rooms["from"].clip(lower=peak_start)
).dt.days.clip(lower=0)
rooms.loc[rooms["full_december"], "surcharge_nights"] = 0
rooms["revenue"] = rooms["nights"] * rooms["Base_Rate"] + rooms["surcharge_nights"] * 50

result = (
    rooms.groupby("Room_Type", as_index=False)["revenue"]
    .sum()
    .rename(columns={"revenue": "Dec_Revenue"})
    .sort_values("Room_Type")
    .reset_index(drop=True)
)

print(result)
print("Matches expected:", result.equals(expected))
